lib/api/src/sys/imports.rs

```diff
@@ -0,0 +1,343 @@
+use std::collections::{HashMap, VecDeque};
+use std::fmt;
+use std::sync::{Arc, Mutex};
+use wasmer_engine::{Export, ImportError, LinkError};
+
+/// TODO add doc
+#[derive(Clone, Default)]
+pub struct Imports {
+    map: Arc<Mutex<HashMap<(String, String), Extern>>>,
+}
+
+impl Imports {
+    /// Create a new `Imports`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Gets an export given a module and a name
+    ///
+    /// # Usage
+    /// ```ignore
+    /// let mut import_object = Imports::new();
+    /// import_object.get_export("module", "name");
+    /// ```
+    pub fn get_export(&self, module: &str, name: &str) -> Option<Export> {
+        let guard = self.map.lock().unwrap();
+        let map_ref = guard.borrow();
+        if map_ref.contains_key(&(module.to_string(), name.to_string())) {
+            let ext = &map_ref[&(module.to_string(), name.to_string())];
+            return Some(ext.to_export());
+        }
+        None
+    }
+
+    /// Returns true if the Imports contains namespace with the provided name.
+    pub fn contains_namespace(&self, name: &str) -> bool {
+        self.map
+            .lock()
+            .unwrap()
+            .borrow()
+            .keys()
+            .any(|(k, _)| (k == name))
+    }
+
+    /// Register a list of externs into a namespace.
+    ///
+    /// # Usage:
+    /// ```ignore
+    /// # use wasmer::{Imports, Exports};
+    /// let mut exports = Exports::new()
+    /// exports.insert("memory", memory);
+    ///
+    /// let mut import_object = Imports::new();
+    /// import_object.register("env", exports);
+    /// // ...
+    /// ```
+    pub fn register_namespace(
+        &mut self,
+        ns: &str,
+        contents: impl IntoIterator<Item = (String, Extern)>,
+    ) {
+        let mut guard = self.map.lock().unwrap();
+        let map = guard.borrow_mut();
+
+        for (name, extern_) in contents.into_iter() {
+            map.insert((ns.to_string(), name.clone()), extern_);
+        }
+    }
+
+    /// TODO: Add doc
+    pub fn define(&mut self, ns: &str, name: &str, val: impl Into<Extern>) {
+        let mut guard = self.map.lock().unwrap();
+        let map = guard.borrow_mut();
+        map.insert((ns.to_string(), name.to_string()), val.into());
+    }
+
+    /// Returns the contents of a namespace as an `Exports`.
+    ///
+    /// Returns `None` if the namespace doesn't exist.
+    pub fn get_namespace_exports(&self, name: &str) -> Option<Exports> {
+        let map = self.map.lock().unwrap();
+        let ret: Exports = map
+            .iter()
+            .filter(|((ns, _), _)| ns == name)
+            .map(|((_, name), e)| (name.clone(), e.clone()))
+            .collect();
+        if ret.is_empty() {
+            None
+        } else {
+            Some(ret)
+        }
+    }
+
+    /// Returns a list of all externs defined in all namespaces.
+    pub fn externs_vec(&self) -> Vec<(String, String, Extern)> {
```

Comment:
I'd remove this method. If you need access to the contents of an `Imports` then you should use the iterator interface.